### ISSUE TITLE
Does sonic-moe plan to support mxfp4 w4a16?

### ISSUE THREAD

**Issue by yuan-luo** [Dao-AILab/sonic-moe#53]
DSV4 support mxfp4 moe weights, does sonic-moe plan to follow up?

**Comment by tridao** [comment]
Eventually, but we'll do mxfp8 first. For now if you want MXFP4, deepgemm is the way to go.

**Comment by yuan-luo** [comment]
> Eventually, but we'll do mxfp8 first. For now if you want MXFP4, deepgemm is the way to go.

Thanks @tridao , I forgot to mention we are using hopper, deepgemm doesn't support mxfp4 in hopper, it only supports sm100 fp8_fp4 gemm.

| Function | SM90 (Hopper) | SM100 (Blackwell) |
|---|---|---|
| FP8×FP8 GEMM | ✅ `sm90_fp8_gemm_1d1d/1d2d` | ✅ `sm100_fp8_gemm_1d1d` |
| FP8×FP4 GEMM | ❌ | ✅ `sm100_fp8_fp4_gemm_1d1d` |
| FP4 MQA/Attention | ❌ | ✅ `sm100_fp4_mqa_logits` |
| Mega MoE (fused EP) | ❌ | ✅ `sm100_fp8_fp4_mega_moe` |
| Grouped GEMM (MoE) | ✅ FP8 only (`MGrouped`) | ✅ FP8 + FP4 |

**Comment by tridao** [comment]
i don't think mxfp4 is supported on hopper hardware

**Comment by yuan-luo** [comment]
> i don't think mxfp4 is supported on hopper hardware

Hopper doesn't have tcgen05.mma, it doesn't support the native mxfp4 processing as on blackwell. The library like humming supports mxfp4 dequant to bf16, providing w4a16 moe mechanism on hopper.

**Comment by tridao** [comment]
Is the dequant done in a separate kernel or inside the moe kernel?

**Comment by yuan-luo** [comment]
> Is the dequant done in a separate kernel or inside the moe kernel?

@tridao Sorry for the late reply, I recently ported a mxfp4 w4a16 flashinfer cutlass moe kernel to sglang for hopper, which is LUT-based FP4→BF16 dequant conversion inside the moe kernel. 
sglang pr: https://github.com/sgl-project/sglang/pull/24816
flashinfer pr: https://github.com/flashinfer-ai/flashinfer/pull/3084

If sonie-moe could implement this kind of kernel for hopper, that would be awesome.

**Comment by yuan-luo** [comment]
Digging further the original implementation is this PR.
https://github.com/flashinfer-ai/flashinfer/pull/1396

**Comment by yuan-luo** [comment]
@tridao I know it's not related with this issue, but could you take a look at this PR? I observed accuracy loss in long output for RelaySSM.
https://github.com/sgl-project/sglang/pull/28695
